hawkular-wildfly-agent/src/main/java/org/hawkular/agent/monitor/protocol/EndpointService.java

```diff
@@ -226,9 +244,14 @@ public ProtocolDiagnostics getDiagnostics() {
      * @param listener to add
      */
     public void addInventoryListener(InventoryListener listener) {
-        status.assertInitialOrStopped(getClass(), "addInventoryListener()");
-        this.inventoryListenerSupport.inventoryListeners.add(listener);
-        LOG.debugf("Added inventory listener [%s] for endpoint [%s]", listener, getMonitoredEndpoint());
+        try {
+            this.inventoryListenerSupport.inventoryListenerRWLock.writeLock().lock();
```

Comment:
acquisition of all locks should always come outside (just before) entering the try-catch. Because if something goes wrong, the finally block is going to try to unlock a lock you have not acquired. So swap lines 247-248. Do the same for the others.